### ISSUE TITLE
Detector de Rupturas  

### ISSUE THREAD

**Issue by edduchzsant21-prog** [edduchzsant21-prog/estructura.-geom-trica.-en-Series.-financieras#1]
# ==========================================================
# Motor de Auditoría de Rupturas — versión WALK-FORWARD
# (corrige el look-ahead bias del PCA/StandardScaler original)
# ==========================================================
#
# DIFERENCIA CLAVE respecto a la versión anterior:
# En vez de hacer fit_transform() una sola vez sobre TODA la serie
# (lo cual usa información futura para calcular el PC1 de fechas pasadas),
# acá el scaler y el PCA se re-calibran en una ventana rodante que
# SOLO usa datos hasta el día t. El PC1 del día t+1 se calcula
# aplicando (transform, no fit) los parámetros aprendidos hasta t.
#
# Esto es más lento y más "feo" en el código, pero es la única forma
# honesta de simular lo que un trader real hubiera visto en tiempo real.

!pip install yfinance scikit-learn scipy matplotlib pandas numpy -q

import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import yfinance as yf
import matplotlib.pyplot as plt

from scipy.signal import hilbert
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

class MotorAuditoriaRupturasWF:
    """
    Versión walk-forward: el PCA y el StandardScaler se re-ajustan
    en cada paso usando SOLO datos pasados (ventana de calibración),
    y se aplican (transform) al dato nuevo. Nunca miran el futuro.
    """

    def __init__(
        self,
        horizontes=None,
        ventana_zscore=252,
        ventana_calibracion=504,   # ~2 años de historia para ajustar PCA/scaler
        paso_recalibracion=21      # recalibrar cada 21 días hábiles (~1 mes), no todos los días
    ):
        self.horizontes = horizontes or [5, 10, 21, 63]
        self.ventana_zscore = ventana_zscore
        self.ventana_calibracion = ventana_calibracion
        self.paso_recalibracion = paso_recalibracion

    def generar_caracteristicas(self, precios):
        precios = precios.dropna().astype(float)
        retornos = np.log(precios / precios.shift(1))

        df_features = pd.DataFrame(index=precios.index)
        for h in self.horizontes:
            df_features[f"ret_{h}"] = retornos.rolling(window=h).sum()
            df_features[f"vol_{h}"] = retornos.rolling(window=h).std()

        return df_features.dropna()

    def detectar_rupturas_walkforward(self, df_features, precios):
        """
        Recorre la serie en el tiempo. Cada `paso_recalibracion` días,
        re-ajusta scaler+PCA con los últimos `ventana_calibracion` días
        (sin incluir el día actual ni futuros). Aplica ese ajuste a los
        días siguientes hasta la próxima recalibración.
        """
        n = len(df_features)
        min_inicio = self.ventana_calibracion

        if n <= min_inicio:
            raise ValueError(
                f"Se necesitan más de {min_inicio} observaciones para calibrar; "
                f"hay {n}."
            )

        pc1_valores = np.full(n, np.nan)
        columnas = df_features.columns
        X_all = df_features[columnas].values

        scaler = None
        pca = None

        for t in range(min_inicio, n):
            # Recalibrar cada `paso_recalibracion` días (o en el primer punto válido)
            if (t - min_inicio) % self.paso_recalibracion == 0 or scaler is None:
                # Ventana de calibración: SOLO pasado, termina en t-1 (no incluye t)
                X_calib = X_all[t - self.ventana_calibracion: t]

                scaler = StandardScaler()
                X_calib_scaled = scaler.fit_transform(X_calib)

                pca = PCA(n_components=1)
                pca.fit(X_calib_scaled)

                # Fijar signo del PC1 de forma determinística
                # (PCA puede voltear el signo entre recalibraciones; lo anclamos
                # a que el primer loading de 'ret_' más corto sea positivo)
                primer_ret_col = [c for c in columnas if c.startswith("ret_")][0]
                idx_col = list(columnas).index(primer_ret_col)
                if pca.components_[0, idx_col] < 0:
                    pca.components_ = -pca.components_

            # Aplicar (transform) el ajuste vigente al dato del día t
            x_t = X_all[t:t + 1]
            x_t_scaled = scaler.transform(x_t)
            pc1_t = pca.transform(x_t_scaled).ravel()[0]
            pc1_valores[t] = pc1_t

        df_resultados = pd.DataFrame(index=df_features.index)
        df_resultados["PC1"] = pc1_valores
        df_resultados["Precio"] = precios.reindex(df_resultados.index)
        df_resultados = df_resultados.dropna(subset=["PC1"])

        # Z-score móvil del PC1 (esto ya era walk-forward correcto en el original,
        # porque rolling().mean()/std() solo usa el pasado)
        media_movil = df_resultados["PC1"].rolling(window=self.ventana_zscore).mean()
        std_movil = df_resultados["PC1"].rolling(window=self.ventana_zscore).std()

        df_resultados["Z_score"] = (
            (df_resultados["PC1"] - media_movil) / std_movil
        )

        return df_resultados.dropna()

    @staticmethod
    def calcular_entropia_shannon(serie_retornos, n_bins=10):
        x = serie_retornos.dropna().values
        if len(x) < 2:
            return np.nan
        hist, _ = np.histogram(x, bins=n_bins, density=False)
        p = hist / hist.sum()
        p = p[p > 0]
        return float(-np.sum(p * np.log2(p)))

    @staticmethod
    def calcular_entropia_hugo(serie_retornos, n_bins=12):
        x = serie_retornos.dropna().values
        if len(x) < 4:
            return np.nan

        x_centrado = x - np.mean(x)
        señal_analitica = hilbert(x_centrado)
        fase = np.angle(señal_analitica)
        theta = np.mod(fase, 2 * np.pi)

        hist, _ = np.histogram(theta, bins=n_bins, range=(0, 2 * np.pi), density=False)
        p = hist / hist.sum()
        p = p[p > 0]

        entropia = -np.sum(p * np.log2(p))
        return float(entropia / np.log2(n_bins))

    @staticmethod
    def filtrar_eventos(eventos, dias_exclusion=10):
        if eventos.empty:
            return eventos.copy()

        eventos = eventos.copy()
        eventos["abs_z"] = eventos["Z_score"].abs()
        eventos = eventos.sort_values("abs_z", ascending=False)

        fechas_seleccionadas = []
        for fecha in eventos.index:
            hay_evento_cercano = any(
                abs((fecha - fecha_anterior).days) <= dias_exclusion
                for fecha_anterior in fechas_seleccionadas
            )
            if not hay_evento_cercano:
                fechas_seleccionadas.append(fecha)

        return eventos.loc[fechas_seleccionadas].sort_index().drop(columns=["abs_z"])

    def auditar_entorno_eventos(
        self, precios, df_resultados, umbral=2.58,
        ventana_dias=21, dias_exclusion=10,
        bins_shannon=10, bins_hugo=12
    ):
        precios = precios.dropna().astype(float)
        retornos = np.log(precios / precios.shift(1)).dropna()

        eventos_brutos = df_resultados[df_resultados["Z_score"].abs() > umbral].copy()
        eventos = self.filtrar_eventos(eventos_brutos, dias_exclusion=dias_exclusion)

        auditorias = []
        for fecha_evento, fila in eventos.iterrows():
            if fecha_evento not in retornos.index:
                continue

            posicion = retornos.index.get_loc(fecha_evento)
            retornos_pre = retornos.iloc[posicion - ventana_dias: posicion]
            retornos_post = retornos.iloc[posicion + 1: posicion + 1 + ventana_dias]

            if len(retornos_pre) < ventana_dias or len(retornos_post) < ventana_dias:
                continue

            shannon_pre = self.calcular_entropia_shannon(retornos_pre, n_bins=bins_shannon)
            shannon_post = self.calcular_entropia_shannon(retornos_post, n_bins=bins_shannon)
            hugo_pre = self.calcular_entropia_hugo(retornos_pre, n_bins=bins_hugo)
            hugo_post = self.calcular_entropia_hugo(retornos_post, n_bins=bins_hugo)

            auditorias.append({
                "Fecha_evento": fecha_evento,
                "Precio_evento": float(fila["Precio"]),
                "PC1": float(fila["PC1"]),
                "Z_score": float(fila["Z_score"]),
                "Direccion": "Ruptura positiva" if fila["Z_score"] > 0 else "Ruptura negativa",
                "Shannon_pre": shannon_pre,
                "Shannon_post": shannon_post,
                "Delta_Shannon": shannon_post - shannon_pre,
                "Hugo_pre": hugo_pre,
                "Hugo_post": hugo_post,
                "Delta_Hugo": hugo_post - hugo_pre,
                "Volatilidad_pre": retornos_pre.std(),
                "Volatilidad_post": retornos_post.std(),
                "Delta_volatilidad": retornos_post.std() - retornos_pre.std(),
                "Retorno_pre": retornos_pre.sum(),
                "Retorno_post": retornos_post.sum()
            })

        if not auditorias:
            return pd.DataFrame()

        return pd.DataFrame(auditorias).set_index("Fecha_evento")

    def ejecutar(self, precios, umbral=2.58, ventana_dias=21, dias_exclusion=10):
        features = self.generar_caracteristicas(precios)
        resultados = self.detectar_rupturas_walkforward(
            df_features=features, precios=precios
        )
        auditoria = self.auditar_entorno_eventos(
            precios=precios, df_resultados=resultados,
            umbral=umbral, ventana_dias=ventana_dias,
            dias_exclusion=dias_exclusion
        )
        return features, resultados, auditoria

    @staticmethod
    def graficar_resultados(df_resultados, auditoria, ticker="Activo", umbral=2.58):
        fig, axes = plt.subplots(
            3, 1, figsize=(15, 12), sharex=True,
            gridspec_kw={"height_ratios": [2, 1, 1]}
        )

        axes[0].plot(df_resultados.index, df_resultados["Precio"], color="navy", linewidth=1.2)
        axes[0].set_title(f"{ticker} — Motor de Auditoría de Rupturas (Walk-Forward)")
        axes[0].set_ylabel("Precio")
        axes[0].grid(alpha=0.25)

        axes[1].plot(df_resultados.index, df_resultados["PC1"], color="darkorange", linewidth=1)
        axes[1].axhline(0, color="black", linewidth=0.8)
        axes[1].set_ylabel("PC1 (walk-forward)")
        axes[1].grid(alpha=0.25)

        axes[2].plot(df_resultados.index, df_resultados["Z_score"], color="purple", linewidth=1)
        axes[2].axhline(umbral, color="red", linestyle="--", label=f"+{umbral}")
        axes[2].axhline(-umbral, color="red", linestyle="--", label=f"-{umbral}")
        axes[2].axhline(0, color="black", linewidth=0.8)
        axes[2].set_ylabel("Z-score")
        axes[2].set_xlabel("Fecha")
        axes[2].grid(alpha=0.25)
        axes[2].legend()

        if not auditoria.empty:
            for fecha in auditoria.index:
                for ax in axes:
                    ax.axvline(fecha, color="red", linestyle=":", alpha=0.65)

        plt.tight_layout()
        plt.show()

# ==========================================================
# Ejecución
# ==========================================================

ticker = "BTC-USD"

datos = yf.download(
    ticker, start="2018-01-01", end="2026-08-23",
    auto_adjust=True, progress=False
)
precios = datos["Close"].squeeze().dropna()

motor = MotorAuditoriaRupturasWF(
    horizontes=[5, 10, 21, 63],
    ventana_zscore=252,
    ventana_calibracion=504,   # ~2 años para calibrar PCA/scaler
    paso_recalibracion=21      # recalibrar mensualmente (ajustable)
)

features, resultados, auditoria = motor.ejecutar(
    precios=precios, umbral=2.58, ventana_dias=21, dias_exclusion=10
)

print("Número de eventos auditados (walk-forward):", len(auditoria))

display(
    auditoria[[
        "Precio_evento", "Z_score", "Direccion",
        "Shannon_pre", "Shannon_post", "Delta_Shannon",
        "Hugo_pre", "Hugo_post", "Delta_Hugo",
        "Delta_volatilidad", "Retorno_post"
    ]].round(4)
)

motor.graficar_resultados(
    df_resultados=resultados, auditoria=auditoria,
    ticker=ticker, umbral=2.58
)

auditoria.to_csv("auditoria_entropia_hugo_BTC_walkforward.csv")
print("Archivo guardado: auditoria_entropia_hugo_BTC_walkforward.csv")

# from google.colab import files
# files.download("auditoria_entropia_hugo_BTC_walkforward.csv")

# ==========================================================
# NOTA IMPORTANTE PARA COMPARAR CONTRA LA VERSIÓN ANTERIOR
# ==========================================================
# Correr ambas versiones (original con fit_transform único, y esta
# walk-forward) sobre el MISMO ticker y comparar:
#   1) ¿Cuántos eventos detecta cada una? (probablemente la walk-forward
#      detecta MENOS eventos "limpios" — eso es normal y esperado)
#   2) ¿El Delta_Hugo promedio cambia de signo o pierde significancia?
#   3) Si los resultados walk-forward siguen viéndose bien, es mucho
#      más creíble que reflejen algo real y no un artefacto de look-ahead.
#
# Parámetros para explorar sensibilidad (punto 5 de la lista de tests):
#   - ventana_calibracion: probar 252 (1 año) vs 504 (2 años) vs 756 (3 años)
#   - paso_recalibracion: probar 5, 21, 63 días
